### PR TITLE
feat!: use `BlobTx` instead of `MsgWirePayForBlob` 

### LINKED ISSUE #1098: Stop malleating transactions and remove `MsgWirePayForBlob`
we should implement https://github.com/celestiaorg/celestia-app/pull/1005, which will stop the malleating transactions and instead use the new `BlobTx` type defined in celestia-core.

### PR DESCRIPTION
**Pull request by evan-forbes**
## Overview

reference PR for everything that needs to be changed to remove the wire PFB. This PR **should not** be merged. While there will be at least one rather large PR, many of these changes can be broken up into chunks.

this PR is mega breaking and currently blocked by https://github.com/celestiaorg/celestia-core/pull/893

closes #1098
is the implementation for #1005 

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

File: pkg/shares/share_splitting.go
Comment (rootulp) [inline pkg/shares/share_splitting.go:66]: [nit] this TODO seems resolved per

```go
	shares := make([]Share, 0, data.SquareSize*data.SquareSize)
```

```suggestion
```

**Review by rootulp — COMMENTED** [review]
Awesome work! So excited for this to land

**Comment by evan-forbes** [comment]
I've been trying to isolate things, but since we did everything in one go in celestia-core, this is make it very difficult to get passing tests without redoing a lot of things. will try again tmrw by just removing the code that deletes the wirePFBs, and maybe pull out a testing refactor. We _might_ be able to pull out the estimation code as well

File: app/testutil_test.go
Comment (evan-forbes) [reply app/testutil_test.go:1]: this is a good point, and we can change this.

the thinking was that these test helpers are only used in tests, so I wanted to keep their access to only other local test files. 

this is also partly driven by the primal need to game the test coverage script

File: app/estimate_square_size.go
Comment (evan-forbes) [reply app/estimate_square_size.go:36]: :100: exactly!! 

and yeah the old way feels kinda midtwit-y in that we were trying to pack with the highest possible efficiency at the cost of waaaaayyyy too much complexity (and therefore bugs)

File: x/blob/README.md
Comment (evan-forbes) [reply x/blob/README.md:14]: > since the share commitment isn't in the MsgPayForBlob or MsgPayForBlobTx

isn't the share commitment in the MsgPayForBlob tho? 

we can get rid of the word "original" tho

File: app/checkTx.go
Comment (evan-forbes) [inline app/checkTx.go:43]: [for reviewers]
here we are parsing the blob transactions before we pass them to state machine. Note that `ProcessBlobTx` is performing the important checks that used to be performed by the `ValidateBasic` method of `MsgWirePFB`.

File: app/estimate_square_size.go
Comment (evan-forbes) [inline app/estimate_square_size.go:24]: [for reviewers]

the estimation process has been significantly simplified by overestimating the number of compact shares used, and always using that estimate as the start of the sparse shares.

File: app/test/fuzz_abci_test.go
Comment (evan-forbes) [inline app/test/fuzz_abci_test.go:52]: I've ran this locally for roughly an hour with no errors. Due to the simpler square estimation, square writing, and removal of the malleation process, I'm extremely hopeful that we will have a meaningful improvement in stability.j

however, due to how the transactions are generated and relatively large amount of compute required for this, the test does take a while. we might want to consider only running this test when we are not using the short flag.

File: proto/blob/tx.proto
Comment (evan-forbes) [inline proto/blob/tx.proto:33]: :tada:

File: testutil/blobfactory/payforblob_factory.go
Comment (evan-forbes) [inline testutil/blobfactory/payforblob_factory.go:1]: we could probably still use a refactor our to our testing utilities, but these I think this is a step in the right direction.

File: app/write_square.go
Comment (evan-forbes) [inline app/write_square.go:97]: first code from gpt chat bot in celestia-app :robot:

File: app/write_square.go
Comment (evan-forbes) [inline app/write_square.go:62]: [for reviewers] 

ideally we should only be hitting this if we're at the max squaresize. that might be something we want to add a test for in the future

File: app/write_square.go
Comment (evan-forbes) [inline app/write_square.go:43]: this is our replacement for the malleation process

File: x/blob/types/blob_tx.go
Comment (evan-forbes) [inline x/blob/types/blob_tx.go:49]: [for reviewers]
this function verifies that the blob transaction is valid. Notably, it performs important check that otherwise would have been performed in the `ValidateBasic` method of `MsgWirePFB`. It's very important that we call this function during check tx and prepare proposal

**Review by evan-forbes — COMMENTED** [review]
okayyyy so I tried to split it up and ended up just writing a few more unit tests instead after failing to do so without causing a bunch of tests to fail :shrug: this is mainly due to the fact that we also did everything at once in core. 

the good news is this PR massively simplifies the application, so hopefully that makes its easier to review than the last mega PR... We are removing the malleation step entirely, using a slightly less efficient by waaaayyyy simpler method of creating the square layout/writing the square. 

The rest of the changes are mainly creating a bunch of testing utility code and ofc deleting `MsgWirePFB`

File: app/estimate_square_size.go
Comment (rootulp) [reply app/estimate_square_size.go:35]: > at the cost of waaaaayyyy too much complexity (and therefore bugs)

+1

File: x/blob/README.md
Comment (rootulp) [reply x/blob/README.md:14]: Sorry I confused myself. You're right b/c the `MsgPayForBlob` contains the share commitment so this was correct as-is 🤦‍♂️

File: app/parse_txs.go
Comment (rootulp) [inline app/parse_txs.go:13]: [optional] two helpers that may make usage of parsedTx slightly more readable

```go
func (p parsedTx) isNormalTx() bool {
	return len(p.normalTx) != 0
}

func (p parsedTx) isBlobTx() bool {
	return len(p.blobTx.Tx) != 0
}
```

File: app/test/check_tx_test.go
Comment (rootulp) [inline app/test/check_tx_test.go:68]: [typo]

```suggestion
			name:      "invalid transaction, mismatched namespace",
```

File: app/test/fuzz_abci_test.go
Comment (rootulp) [reply app/test/fuzz_abci_test.go:52]: > when we are not using the short flag.

+1

Related https://github.com/celestiaorg/celestia-app/issues/1025

File: app/testutil_test.go
Comment (rootulp) [reply app/testutil_test.go:1]: > this is also partly driven by the primal need to game the test coverage script

😆 

I'm not strongly opinionated on this so defer to you

File: app/write_square.go
Comment (rootulp) [reply app/write_square.go:97]: our jobs will be obsolete soon 😨

File: app/write_square_test.go
Comment (rootulp) [inline app/write_square_test.go:71]: [question] is this test verifying that a square size of 4 can't contain a 10,000 byte blob because 10,000 / 503 (sparse share content size) is 19.88071571 and only 16 shares exist in a square size of 4?

File: app/write_square_test.go
Comment (rootulp) [inline app/write_square_test.go:89]: [no change needed] nice test 👍 verifies that blobs are sorted by namepsace

File: go.mod
Comment (rootulp) [inline go.mod:185]: [optional] I assume the snapshot version was used so that PRs could be published simultaneously but reminder that we likely want to merge https://github.com/celestiaorg/celestia-core/pull/893 , cut a release, and use that version here

File: proto/blob/tx.proto
Comment (rootulp) [reply proto/blob/tx.proto:33]: 🪓

File: x/blob/README.md
Comment (rootulp) [inline x/blob/README.md:89]: `wblob` appears undefined

File: x/blob/types/test/blob_tx_test.go
Comment (rootulp) [inline x/blob/types/test/blob_tx_test.go:55]:
```suggestion
			name: "invalid transaction, mismatched namespace",
```

File: x/blob/types/test/blob_tx_test.go
Comment (rootulp) [inline x/blob/types/test/blob_tx_test.go:65]:
```suggestion
			name: "invalid transaction, no pfb",
```

**Review by rootulp — APPROVED** [review]
Incredible work @evan-forbes ! Very net negative

Screenshot 2022-12-08 at 5 37 58 PM

I don't have any blocking feedback

File: x/blob/types/blob_tx.go
Comment (cmwaters) [inline x/blob/types/blob_tx.go:45]: So with this, it's already possible to submit multiple other random sdk.Msg's in the `BlobTx` and have multiple PFBs/Blobs?

File: x/blob/types/blob_tx.go
Comment (cmwaters) [inline x/blob/types/blob_tx.go:57]: Shouldn't this error? If this is not true isn't this an incorrectly constructed `BlobTx`?

File: x/blob/types/blob_tx.go
Comment (cmwaters) [inline x/blob/types/blob_tx.go:96]: why not do this now :) ?

File: x/blob/types/blob_tx.go
Comment (cmwaters) [inline x/blob/types/blob_tx.go:112]: Don't we allow multiple blobs per `Tx` now?

File: x/blob/types/payforblob.go
Comment (cmwaters) [inline x/blob/types/payforblob.go:31]: Is `commitment` a fixed size? (like a hash)

File: app/estimate_square_size.go
Comment (evan-forbes) [reply app/estimate_square_size.go:49]: [compact shares](https://pkg.go.dev/github.com/celestiaorg/celestia-app@v0.10.0/pkg/shares#hdr-Compact_Share_Schema) are shares that can contain multiple individual things in each one. This encoding scheme allows for increased efficiency as to what can fit in those shares. Since we've removed evidence from the square, this is only celestia-specific transactions, but could include other things in the future such as ISRs (intermediate state roots). 

This is opposed to Sparse shares, which only contain a single thing in those shares. This is how blobs are encoded.

File: app/estimate_square_size.go
Comment (evan-forbes) [reply app/estimate_square_size.go:37]: #1101

File: app/write_square_test.go
Comment (evan-forbes) [reply app/write_square_test.go:71]: yeah exactly

File: x/blob/README.md
Comment (evan-forbes) [reply x/blob/README.md]: incorporated the original suggestion to just remove this blurb of code. Instead, I left links to celestia's nodes submission docs and pointed to the function to reverse engineer for programmatic usage.

7f3148d264c658550c6ad0e76c7561215ec1093e

File: x/blob/types/blob_tx.go
Comment (evan-forbes) [reply x/blob/types/blob_tx.go:96]: ha! whoops good catch. We already had an sdk error for this, I probably just forgot to while writing this. 

c6538d6e4d2a6c15426e862a1ac5a4015442bbd6

File: x/blob/types/blob_tx.go
Comment (evan-forbes) [reply x/blob/types/blob_tx.go]: yes, great point. we can and should reject all transaction that one or more invalid PFBs. 

fixed here 4d712c9e21ccaea81531e8d52cd34305b62fefd5

File: x/blob/types/blob_tx.go
Comment (evan-forbes) [reply x/blob/types/blob_tx.go:112]: we do not currently have that functionality, but we will soon #388

File: x/blob/types/payforblob.go
Comment (rootulp) [reply x/blob/types/payforblob.go:31]: Yes `commitment` is a fixed size hash

Ref:
1. https://github.com/celestiaorg/celestia-app/blob/a82110a281bf9ee95a9bf9f0492e5d091371ff0b/x/blob/types/payforblob.go#L164
2. https://github.com/tendermint/tendermint/blob/ab4238a0e27acff1cb4fe41cc4b1dcebbd706961/crypto/merkle/tree.go#L9-L13

File: app/checkTx.go
Comment (evan-forbes) [reply app/checkTx.go]: good catch idek why I named it that, the test file is named normally lol

https://github.com/celestiaorg/celestia-app/pull/1089/commits/f1d01d343dfac07ce5d39e680214f2abbc306cef

File: app/checkTx.go
Comment (evan-forbes) [reply app/checkTx.go]: good idea https://github.com/celestiaorg/celestia-app/pull/1089/commits/4ed9f50cdc51efc78114c9f7d4140a3449150cd4

File: app/parse_txs.go
Comment (evan-forbes) [reply app/parse_txs.go:12]: cool, and yeah good idea, added for a future PR.

File: x/blob/types/blob_tx.go
Comment (evan-forbes) [reply x/blob/types/blob_tx.go:45]: that's a good question, and I did some testing and it wasn't working. 

That wasn't an explicit goal of this PR, so instead of leaving ambiguous I added an extra temporary explicit check to prevent this. https://github.com/celestiaorg/celestia-app/pull/1089/commits/5978eef3c6b3cb7a4beb9d89a6081735e5f84c17

Note that this additional check does not change actually enforce that as a block validity rule! Its only a default. If a block producer does this now, the transaction will fail, but pfb and its blob will still get included. This is not unlike submitting a pfb from an account that doesn't have enough funds to pay for gas, the blob will still be included. https://github.com/celestiaorg/celestia-app/issues/890#issuecomment-1344984604

File: app/test/block_size_test.go
Comment (cmwaters) [inline app/test/block_size_test.go:101]:
```suggestion
	// generate 80 randomly sized txs (max size == 100kb) by generating these
```

File: testutil/test_app.go
Comment (cmwaters) [inline testutil/test_app.go:184]: Why remove `t *testing.T`?

File: x/blob/types/payforblob.go
Comment (cmwaters) [reply x/blob/types/payforblob.go:31]: Ok so we could assert a specific size rather than non-zero in `ValidateBasic`:

```
if len(msg.ShareCommitment) == 0 {
	return ErrEmptyShareCommitment
}
```

File: x/blob/README.md
Comment (cmwaters) [inline x/blob/README.md:49]: Is this up to date? It takes me to the removed `MsgWirePayForBlob`

File: x/blob/README.md
Comment (cmwaters) [inline x/blob/README.md:47]: So this means you can't submit it directly but have to proxy it through the "halo" network

**Review by cmwaters — APPROVED** [review]
Looks good from my angle but more reviews would be great

File: x/blob/README.md
Comment (evan-forbes) [reply x/blob/README.md:47]: transactions can be submitted via all normal sdk/tendermint mechanisms, and the light client network simply wraps those endpoints. its a good point tho, and we should make that clearer in our docs

https://github.com/celestiaorg/celestia-app/issues/1113

File: x/blob/README.md
Comment (evan-forbes) [reply x/blob/README.md:49]: this is not update to date, but I think its something that we will have to handle in #1113 after we merge this, as the most up to date reference will be dead after this branch gets deleted

https://github.com/celestiaorg/celestia-app/issues/1113#issuecomment-1347726776

File: testutil/test_app.go
Comment (evan-forbes) [reply testutil/test_app.go:184]: tbh, I don't remember fully... man this PR really blew up in scope... at once point in time, I think I was using this, or some util that used this function somewhere that didn't have access, such as the testnode package, but I don't think those changes made it to this commit.

the good news is that errors that occur in this function would have done something similar when requiring that there be no errors. We should still fix this in https://github.com/celestiaorg/celestia-app/issues/1114

File: x/blob/README.md
Comment (evan-forbes) [reply x/blob/README.md:47]: kinda related https://github.com/celestiaorg/celestia-app/pull/1089#discussion_r1044708058

**Review by rootulp — APPROVED** [review]
:shipit:

**Pull request merged**
